File: 0332.ReconstructItinerary.py
from collections import defaultdict

class Solution(object):
    def findItinerary(self, tickets):
        """
        :type tickets: List[List[str]]
        :rtype: List[str]
        """
        targets = {}
        for f, t in sorted(tickets)[::-1]:
            targets[f] = targets.get(f, []) + [t]
        path, stack = [], ['JFK']
        while stack:
            while targets.get(stack[-1]):
                stack.append(targets[stack[-1]].pop())
            path.append(stack.pop())
        '''
        self.targets = defaultdict(list)
        for f, t in sorted(tickets)[::-1]:
            self.targets[f].append(t)
        path = []
        self.dfs('JFK', path)
        '''
        return path[::-1]

    def dfs(self, airport, path):
        while self.targets[airport]:
            self.dfs(self.targets[airport].pop(), path)
        path.append(airport)

File: 0332.ReconstructItinerary.go
func findItinerary(tickets [][]string) []string {
    Map := make(map[string][]string)
    for _, t := range tickets {
        Map[t[0]] = append(Map[t[0]], t[1])
    }
    for _, to := range Map {
        sort.Strings(to)
    }
    path := make([]string, 0, len(tickets)+1)
    stack := make([]string, len(tickets)+1)
    stack[0] = "JFK"
    n := 1
    for n > 0 {
        for len(Map[stack[n-1]]) > 0 {
            cur := stack[n-1]
            stack[n] = Map[cur][0]
            n++
            Map[cur] = Map[cur][1:]
        }
        path = append([]string{stack[n-1]}, path...)
        n--
    }
    return path
}